library/kani/src/arbitrary.rs

```diff
@@ -1,33 +1,103 @@
 /// its type.
 pub trait Arbitrary {
     fn any() -> Self;
 }
 
-impl<T> Arbitrary for T
-where
-    T: Invariant,
-{
-    default fn any() -> Self {
-        let value = unsafe { any_raw::<T>() };
-        assume(value.is_valid());
-        value
+/// The given type can be represented by an unconstrained symbolic value of size_of::<T>.
+macro_rules! trivial_arbitrary {
+    ( $type: ty ) => {
+        impl Arbitrary for $type {
+            #[inline(always)]
+            fn any() -> Self {
+                unsafe { crate::any_raw_internal::<$type>() }
+            }
+        }
+    };
+}
+
+trivial_arbitrary!(u8);
+trivial_arbitrary!(u16);
+trivial_arbitrary!(u32);
+trivial_arbitrary!(u64);
+trivial_arbitrary!(u128);
+trivial_arbitrary!(usize);
+
+trivial_arbitrary!(i8);
+trivial_arbitrary!(i16);
+trivial_arbitrary!(i32);
+trivial_arbitrary!(i64);
+trivial_arbitrary!(i128);
+trivial_arbitrary!(isize);
+
+// We do not constraint floating points values per type spec. Users must add assumptions to their
+// verification code if they want to eliminate NaN, infinite, or subnormal.
+trivial_arbitrary!(f32);
+trivial_arbitrary!(f64);
+
+trivial_arbitrary!(());
+
+impl Arbitrary for bool {
+    #[inline(always)]
+    fn any() -> Self {
+        let byte = u8::any();
+        byte == 0
     }
 }
 
+/// Validate that a char is not outside the ranges [0x0, 0xD7FF] and [0xE000, 0x10FFFF]
+/// Ref: <https://doc.rust-lang.org/stable/nomicon/what-unsafe-does.html>
+impl Arbitrary for char {
+    #[inline(always)]
+    fn any() -> Self {
+        // Generate an arbitrary u32 and constrain it to make it a valid representation of char.
+        let val = u32::any();
+        crate::assume(val <= 0xD7FF || (0xE000..=0x10FFFF).contains(&val));
+        unsafe { char::from_u32_unchecked(val) }
+    }
+}
+
+macro_rules! nonzero_arbitrary {
+    ( $type: ty, $base: ty ) => {
+        impl Arbitrary for $type {
+            #[inline(always)]
+            fn any() -> Self {
+                let val = <$base>::any();
+                crate::assume(val != 0);
+                unsafe { <$type>::new_unchecked(val) }
+            }
+        }
+    };
+}
+
+nonzero_arbitrary!(NonZeroU8, u8);
+nonzero_arbitrary!(NonZeroU16, u16);
+nonzero_arbitrary!(NonZeroU32, u32);
+nonzero_arbitrary!(NonZeroU64, u64);
+nonzero_arbitrary!(NonZeroU128, u128);
+nonzero_arbitrary!(NonZeroUsize, usize);
+
+nonzero_arbitrary!(NonZeroI8, i8);
+nonzero_arbitrary!(NonZeroI16, i16);
+nonzero_arbitrary!(NonZeroI32, i32);
+nonzero_arbitrary!(NonZeroI64, i64);
+nonzero_arbitrary!(NonZeroI128, i128);
+nonzero_arbitrary!(NonZeroIsize, isize);
+
 impl<T, const N: usize> Arbitrary for [T; N]
 where
     T: Arbitrary,
 {
     fn any() -> Self {
         // The "correct way" would be to MaybeUninit but there is performance penalty.
-        let mut data: [T; N] = unsafe { crate::any_raw() };
+        let mut data: [T; N] = unsafe { crate::any_raw_internal() };
```

Comment:
I never fully understood why we call `any_raw_internal` here followed by `any` for every element? Can we instead just initialize the array with zeros and then fill it with nondet data?

Also, from what I understand, this is calling `any_raw_internal` for `[T; N]`. Doesn't this violate the invariant that we only call it on primitives?

Reply:
We can't initialize everything to zero. Depending on the type it would be UB. MaybeUnit is the correct way but the performance was rather bad. Let me see what else I can do.

Reply:
Would this be more efficient:
```rust
[(); N].map(|_| kani::any());
```
?
Source: https://stackoverflow.com/a/69488028

Reply:
I can give it a shot. Thanks

Reply:
Yep, it's slightly faster and much cleaner. Thanks for the suggestion!